src/contract_spec.ts

```diff
@@ -0,0 +1,643 @@
+          } else {
+            // otherwise ensure that the types of the values are all the same
+            if (val.length > 0 && val.some((v) => typeof v !== typeof val[0])) {
+              throw new TypeError(
+                `array values (${val}) must have the same type (types: ${val
+                  .map((v) => typeof v)
+                  .join(",")})`
+              );
+            }
+          }
+          return xdr.ScVal.scvVec(
+            val.map((v) => this.nativeToScVal(v, elementType))
+          );
+        }
+        if (val.constructor === Map) {
+          if (typeDef?.switch().name !== "scSpecTypeMap") {
+            throw new TypeError(
+              `Type ${typeDef} was not map, but value was Map`
+            );
+          }
+          let scMap = typeDef!.value() as xdr.ScSpecTypeMap;
+          let map = val as Map<any, any>;
+          let entries: xdr.ScMapEntry[] = [];
+          map.forEach(([k, v]) => {
+            let key = this.nativeToScVal(k, scMap.keyType());
+            let val = this.nativeToScVal(v, scMap.valueType());
+            entries.push(new xdr.ScMapEntry({ key, val }));
+          });
+          return xdr.ScVal.scvMap(entries);
+        }
+
+        if ((val.constructor?.name ?? "") !== "Object") {
+          throw new TypeError(
+            `cannot interpret ${
+              val.constructor?.name
+            } value as ScVal (${JSON.stringify(val)})`
+          );
+        }
+
+        return xdr.ScVal.scvMap(
+          Object.entries(val).map(([k, v]) => {
+            // the type can be specified with an entry for the key and the value,
+            // e.g. val = { 'hello': 1 } and opts.type = { hello: [ 'symbol',
+            // 'u128' ]} or you can use `undefined` for the default interpretation
+            return new xdr.ScMapEntry({
+              key: this.nativeToScVal(k, undefined),
+              val: this.nativeToScVal(v, undefined),
+            });
+          })
+        );
+      }
+
+      case "number":
+      case "bigint": {
+        let ty = typeDef ?? xdr.ScSpecTypeDef.scSpecTypeU32();
+        switch (ty.switch().name) {
+          case "scSpecTypeU32":
+            return xdr.ScVal.scvU32(val as number);
+          case "scSpecTypeI32":
+            return xdr.ScVal.scvI32(val as number);
+          case "scSpecTypeU64":
+            return new ScInt(val, { type: "u64" }).toU64();
+          case "scSpecTypeI64":
+            return new ScInt(val, { type: "i64" }).toI64();
+          case "scSpecTypeU128":
+            return new ScInt(val, { type: "u128" }).toU128();
+          case "scSpecTypeI128":
+            return new ScInt(val, { type: "i128" }).toI128();
+          case "scSpecTypeU256":
+            return new ScInt(val, { type: "u256" }).toU256();
+          case "scSpecTypeI256":
+            return new ScInt(val, { type: "i256" }).toI256();
+          default:
+            throw new TypeError(`invalid type (${ty}) specified for integer`);
```

Comment:
suggestion to significantly simplify this: 
```typescript
// convert well-formed spec type to int type, e.g. 'scSpecTypeU256' --> 'u256'
const intType = ty.switch().name.substring(10).toLowerCase();
return new XdrLargeInt(intType, val as number).toScVal();
```

`XdrLargeInt` will do the error-handling, or you can try-catch if you want to bubble up `ty` specifically.

Reply:
Will do!

Reply:
Also wouldn't I want `as bigint` ?

Reply:
Bah, that's a typescript bug then - the js itself has a `toScVal()`: https://github.com/stellar/js-stellar-base/blob/soroban/src/numbers/xdr_large_int.js#L202, so you can `@ts-ignore` for now. As for `bigint`, I think you're right - I didn't read into that part closely and just borrowed the `as number` you had.

Reply:
Bah, that's a typescript bug then - the js itself has a `toScVal()`: https://github.com/stellar/js-stellar-base/blob/soroban/src/numbers/xdr_large_int.js#L202, so you can `@ts-ignore` for now. As for `bigint`, I think you're right - I didn't read into that part closely and just borrowed the `as number` you had.

Reply:
This is fixed in the latest stellar-base!